sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/CostBasedJoinReorder.scala

```diff
@@ -150,9 +150,16 @@ object JoinReorderDP extends PredicateHelper with Logging {
     // Level i maintains all found plans for i + 1 items.
     // Create the initial plans: each plan is a single item with zero cost.
     val itemIndex = items.zipWithIndex
-    val foundPlans = mutable.Buffer[JoinPlanMap](itemIndex.map {
-      case (item, id) => Set(id) -> JoinPlan(Set(id), item, ExpressionSet(), Cost(0, 0))
-    }.toMap)
+    val foundPlans = mutable.Buffer[JoinPlanMap]({
+      val idToJoinPlanSeq = itemIndex.map {
+        case (item, id) => Set(id) -> JoinPlan(Set(id), item, ExpressionSet(), Cost(0, 0))
+      }
+      // SPARK-32687: Change to use `LinkedHashMap` to make sure that items are
+      // inserted and iterated in the same order.
+      val ret = new mutable.LinkedHashMap[Set[Int], JoinPlan]()
+      idToJoinPlanSeq.foreach(v => ret.put(v._1, v._2))
```

Comment:
I think you can just construct this above, with a .foreach instead of .map? or just call .addAll on the result of the .map.
I guess LinkedHashMap takes more memory, but that shouldn't matter much here.